File: backend/src/shelf/models/standards.py
"""Engineering standards and their revisions.

An engineering standard is republished over time, and which edition
applies is a decision a project makes deliberately. That needs two things
the plain item model can't express: revisions of one standard have to
know they're the same standard, and a space has to be able to say which
of them it uses.

`StandardFamily` is the standard itself (say, "ACME 1234"), independent
of any edition. `StandardRevision` attaches one item to one family — the
item carries the PDF, the title, the metadata; the revision row carries
only what orders it against its siblings.

Families are instance-wide rather than space-scoped on purpose. A given
standard is the same standard whichever space holds a copy, and a project
pinning a revision needs to name a family that the Standards space also
recognises. The trade is that two spaces holding their own copy of the
same edition both point at one family, which is what makes the revision
dropdown work across an inheritance link.
"""

import uuid
from datetime import date, datetime

from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    ForeignKey,
    Index,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import CITEXT
from sqlalchemy.orm import Mapped, mapped_column

from .base import UUIDPK, Base, Timestamps, utcnow


class StandardFamily(UUIDPK, Timestamps, Base):
    """One standard, across every edition of it."""

    __tablename__ = "standard_families"
    __table_args__ = (
        # CITEXT on both halves so a body or designation typed with
        # different capitalisation lands on the same family instead of
        # quietly forking the revision history in two.
        UniqueConstraint("body", "designation", name="uq_standard_families_key"),
    )

    # Whoever publishes it. Free text rather than an enum — the list of
    # bodies an engineering office cares about is long, changes, and
    # isn't ours to fix.
    body: Mapped[str] = mapped_column(CITEXT(), nullable=False)
    # The designation without an edition — "ACME 1234", "ACME-RP-7". The
    # edition lives on the revision.
    designation: Mapped[str] = mapped_column(CITEXT(), nullable=False)
    # Human title, for the picker.
    title: Mapped[str | None] = mapped_column(Text, nullable=True)


class StandardRevision(Base):
    """One item, as one edition of one family.

    `item_id` is the primary key: an item is a revision of at most one
    standard. Deleting the item takes the revision row with it; deleting
    the family is blocked while revisions point at it (RESTRICT), since
    that would silently orphan every pin.
    """

    __tablename__ = "standard_revisions"
    __table_args__ = (
        # The revision dropdown: every sibling of a family, in order.
        Index("ix_standard_revisions_family_issued", "family_id", "issued_on"),
    )

    item_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("items.id", ondelete="CASCADE"), primary_key=True
    )
    family_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("standard_families.id", ondelete="RESTRICT"), nullable=False
    )
    # What this edition is called: "2020", "Rev. 5", "Ed. 3 Amd. 1".
    # Display only — ordering is `issued_on`, because label formats vary
    # by body and don't sort sensibly against each other.
    label: Mapped[str] = mapped_column(String, nullable=False)
    # Publication date. Nullable because a PDF sometimes doesn't say, and
    # a revision with an unknown date is still worth recording — it just
    # sorts last and can never be "latest". Ordering nulls last is what
    # keeps an undated import from displacing a dated current edition.
    issued_on: Mapped[date | None] = mapped_column(Date, nullable=True)
    # Explicitly withdrawn or superseded by the issuing body. Excluded
    # from "latest" regardless of date, so a withdrawn late amendment
    # doesn't get recommended over the current edition.
    superseded: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=utcnow
    )


class SpaceStandardPin(Base):
    """The revision of one family that one space treats as its own.

    A project space pins "we build to the 2018 edition of ACME 1234", and
    its library stops showing the four other editions it inherits from
    the Standards space unless someone asks for them. The pin is per
    space, so the Standards space itself stays the complete record.

    `item_id` is not constrained to the family here — the API checks it,
    because the useful error is "that item isn't a revision of that
    standard" rather than an integrity violation.
    """

    __tablename__ = "space_standard_pins"

    space_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("spaces.id", ondelete="CASCADE"), primary_key=True
    )
    family_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("standard_families.id", ondelete="CASCADE"), primary_key=True
    )
    item_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("items.id", ondelete="CASCADE"), nullable=False
    )
    pinned_by: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=utcnow
    )

File: frontend/src/api/standards.ts
/**
 * Engineering standards: which editions exist, and which one a space
 * uses.
 *
 * A standard is republished over time, and which edition applies is a
 * project decision. Items that are editions of the same standard are
 * linked to one "family", which is what makes the revision dropdown and
 * the latest-edition badge possible. A space pins the revision it builds
 * to; its library then shows that one and hides the siblings unless
 * asked.
 */

import { apiFetch } from "./client";

export interface StandardFamily {
  id: string;
  /** Whoever publishes it. */
  body: string;
  /** Designation without an edition — "ACME 1234", "ACME-RP-7". */
  designation: string;
  title: string | null;
}

export interface StandardRevision {
  item_id: string;
  space_id: string;
  /** What this edition is called: "2020", "Rev. 5". Display only. */
  label: string;
  /** ISO date, or null when the document doesn't say. Undated
   * revisions sort last and are never "latest". */
  issued_on: string | null;
  superseded: boolean;
  title: string | null;
  /** Newest revision *the caller can read*. At most one per response. */
  is_latest: boolean;
  /** The revision the space in the request pins, if any. */
  is_pinned: boolean;
}

export interface RevisionsResponse {
  family: StandardFamily;
  revisions: StandardRevision[];
  /**
   * False when the instance holds a newer edition than any the caller
   * can see. Lets the UI say "there is a newer one, ask for access"
   * rather than presenting a stale edition as current.
   */
  is_latest_known: boolean;
}

/**
 * Every edition of this item's standard that the caller can read.
 *
 * `space` only decides which revision comes back marked pinned;
 * visibility is the caller's either way.
 */
export function fetchRevisions(
  itemId: string,
  space?: string,
): Promise<RevisionsResponse> {
  const qs = space ? `?space=${encodeURIComponent(space)}` : "";
  return apiFetch<RevisionsResponse>(
    `/api/items/${encodeURIComponent(itemId)}/revisions${qs}`,
  );
}

export interface LinkRevisionInput {
  body: string;
  designation: string;
  label: string;
  issued_on?: string | null;
  title?: string | null;
  superseded?: boolean;
}

/**
 * Declare an item to be one edition of one standard.
 *
 * The family is found-or-created from (body, designation), both matched
 * case-insensitively — so a second upload of the same standard joins the
 * existing revision history instead of forking it.
 */
export function linkRevision(
  itemId: string,
  input: LinkRevisionInput,
): Promise<RevisionsResponse> {
  return apiFetch<RevisionsResponse>(
    `/api/items/${encodeURIComponent(itemId)}/revision`,
    { method: "PUT", body: JSON.stringify(input) },
  );
}

export function unlinkRevision(itemId: string): Promise<void> {
  return apiFetch<void>(`/api/items/${encodeURIComponent(itemId)}/revision`, {
    method: "DELETE",
  });
}

// ── Pins ─────────────────────────────────────────────────────────────────────

export interface StandardPin {
  family: StandardFamily;
  item_id: string;
  label: string;
  issued_on: string | null;
}

/** The revisions this space has chosen, one per standard. */
export function fetchPins(slug: string): Promise<StandardPin[]> {
  return apiFetch<StandardPin[]>(
    `/api/spaces/${encodeURIComponent(slug)}/pins`,
  );
}

/**
 * Choose the revision of one standard that this space uses.
 *
 * Owner-only server-side: which edition a project builds to has
 * consequences outside the library, and it changes what everyone else in
 * the space sees by default.
 */
export function setPin(
  slug: string,
  familyId: string,
  itemId: string,
): Promise<StandardPin> {
  return apiFetch<StandardPin>(
    `/api/spaces/${encodeURIComponent(slug)}/pins/${encodeURIComponent(familyId)}`,
    { method: "PUT", body: JSON.stringify({ item_id: itemId }) },
  );
}

export function clearPin(slug: string, familyId: string): Promise<void> {
  return apiFetch<void>(
    `/api/spaces/${encodeURIComponent(slug)}/pins/${encodeURIComponent(familyId)}`,
    { method: "DELETE" },
  );
}
